### ISSUE TITLE
[spec] Potential accessibility issues with singlehtml due to missing MathML

### ISSUE THREAD

**Issue by ngzhian** [WebAssembly/spec#1386]
I was looking at https://github.com/WebAssembly/spec/blob/941c6f37cb13de20f54b89b73145f802b87b8155/document/core/util/mathjax2katex.py#L117

By default, katex will generate something like:
```
<span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mrow><mn mathvariant="monospace">0</mn><mi mathvariant="monospace">x</mi><mrow><mi mathvariant="monospace">F</mi><mi mathvariant="monospace">C</mi></mrow></mrow><mtext> </mtext><mrow><mn mathvariant="monospace">0</mn><mi mathvariant="monospace">x</mi><mn mathvariant="monospace">11</mn></mrow></mrow><annotation encoding="application/x-tex">\mathrm{{\mathtt{0x{FC}}}~{\mathtt{0x{11}}}}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.61111em;vertical-align:0em;"></span><span class="mord"><span class="mord"><span class="mord"><span class="mord mathtt">0x</span><span class="mord"><span class="mord mathtt">FC</span></span></span></span><span class="mspace nobreak"> </span><span class="mord"><span class="mord"><span class="mord mathtt">0x</span><span class="mord"><span class="mord mathtt">11</span></span></span></span></span></span></span></span></span>
```

Notice the `katex-mathml` and `<math>` DOM nodes. We strip those away by skipping directly to `<span class="katex-html">`.

We can can actually skip this post-processing with a [katex option](https://katex.org/docs/options.html#:~:text=output%3A%20string.%20Determines,is%20the%20default.) but it seems like MathML is for accessibility.

I don't now how the singlehtml doc currently is like under, say, a screen reader, and whether the missing MathML is acceptable, but I will go read a bit more about it.

**Comment by ericprud** [comment]
On Monday I'll post a pointer to the output of the sphinxToTr, which I hope to replace the one-pager. It basically packages the sphinx output so if it's in sphinx, it's in this (these) doc

**Comment by ericprud** [comment]
It's one day later than Monday, but here's a snapshot of my plan A for publishing the core spec as a multi-page doc:
  https://www.w3.org/2021/11/wasm-stage/
This is the output of [SphinxToTr](https://github.com/w3c/sphinx-to-tr/). There are a two wasm-specific files in the top dir:

1. [wasm-cfg.yaml](https://github.com/w3c/sphinx-to-tr/blob/main/wasm-cfg.yaml) - pointer to input (Sphinx-generated docs), output, index page, stuff to stick at the top
2. [wasm-respec.html](https://github.com/w3c/sphinx-to-tr/blob/main/wasm-respec.html) - leverage respec to build the index page.

These files would probably go into the core repo if someone anywhere ever wanted to re-use SphinxToTr for another doc.

**Comment by rossberg** [comment]
Thanks @ericprud, that looks promising! How is the table of contents generated? That only seems to pick some index sections from the spec's appendix, not the main TOC. (Also, the links do not seem to work, but I assume that's because it's in staging? Logo is broken as well.)

**Comment by ericprud** [comment]
yeah, forgot the chacl subdirs. may now look promising without having to stretch your generosity so much.

**Comment by ngzhian** [comment]
I'm not familiar with W3C requirements, is there interest in trying to use Sphinx HTML theming system to accomplish what SphinxToTr is doing?

**Comment by ericprud** [comment]
If it can accomplish the same thing with less maintenance, that'd be a win all around.
The mono doc was pretty unwieldy, which was my motivation to create SphinxToTr. It does some stuff that it needs to do
1. document, resource and editor links
2. Status of This Document
3. W3C TR Style
4. W3C-style TOC

, and some stuff that makes it nice to use:
1. TOC on all pages
2. search

**Comment by ngzhian** [comment]
I had a go at trying to so some of this with sphinx/respec:

attempt 1, singlehtml builder with sphinx, then include respec's JS file to generate front-matter and TOC. Problem faced: expanding anchor tags, we have a bunch of `<a href="#something">` where it links to a `<span id="#something">`, and respec is [not happy](https://github.com/w3c/respec/blob/develop/src/core/anchor-expander.js#L46). I think I can workaround this, still trying.

attempt 2, (multi-page) html builder with sphinx, then use [toctree](https://www.sphinx-doc.org/en/master/templating.html?highlight=toc#toctree) to generate the TOC, but can't get numbered TOC because the [toctree](https://github.com/WebAssembly/spec/blob/main/document/core/index.rst) is not defined to be numbered. We can make 2 toctree in index.rst, and use the `..only` directive to only output one toctree, but then the output HTML will contain duplicated TOC tree. It looks like the toctree inside of the templating engine doesn't respect `..only`. Also, this attempt requires us to duplicate the front matter inside sphinx layout (respec takes care of that).

Just fyi.

**Comment by ericprud** [comment]
Attempt 2 sounds encouraging to me, but perhaps that means i fail to grasp the fiddliness of it all. SphinxToTr is also quite fiddly so I'd be happy to work with you to push that logic uphill into Sphinx tools. This would make it less prone to failure when divs get rearranged. I can see numbered TOC being useful outside of our use case. Possibly so would be something that allows you to glue your own front page in.

Am I being too much of a Pollyanna?

**Comment by ngzhian** [comment]
I had a go at attempt 2 again, this is the result: https://www.ngzhian.com/spec/newhtml/ and the changes required are https://github.com/ngzhian/spec/pull/3

Couple of problems:
- toc on the side is duplicated (can be fixed if https://github.com/sphinx-doc/sphinx/pull/9830 is merged upstream)
- w3c css rules don't apply to toc, due to missing classes

**Comment by ericprud** [comment]
We could use a much skinnier script than [SphinxToTr.js](https://github.com/w3c/sphinx-to-tr/) to fix those probs. Do the TOCs work if you manually edit the classes? Can you share the result?

**Comment by rossberg** [comment]
This looks quite promising!

If the Sphinx folks don't get to merge and release that bug fix anytime soon, we could presumably also hack up a little script that pre-processes index.rst appropriately.

**Comment by ngzhian** [comment]
> We could use a much skinnier script than [SphinxToTr.js](https://github.com/w3c/sphinx-to-tr/) to fix those probs. Do the TOCs work if you manually edit the classes? Can you share the result?

It works pretty well, ptal https://www.ngzhian.com/spec/newhtml/. The number of the TOC is tricky, W3C creates separate spans for the numbers, Sphinx toctree has the numbers together with the title. This means we can't get the correct padding/alignment of the TOC lists, but I think it looks fine.

> This looks quite promising!
> 
> If the Sphinx folks don't get to merge and release that bug fix anytime soon, we could presumably also hack up a little script that pre-processes index.rst appropriately.

We can probably do some hacks. Right now the closest I have is to have numbered TOC for both the w3c multi-page (demo-ed here), and also the existing multi-page. Without the patch it is impossible to have 2 different top-level TOC, one numbered, one not numbered.

See multi-page https://www.ngzhian.com/spec/core/
See demo of sphinx-based w3c compatible multi-page  https://www.ngzhian.com/spec/newhtml/

**Comment by ericprud** [comment]
Hi @ngzhian , Can you explain how you generate https://www.ngzhian.com/spec/newhtml/ so I can try to work it into a publication process? @rossberg , @dschuff , @lukewagner , i'll be in SF 24 March - 2 April if it will help to sit in the same room to work out a publication pipeline.

**Comment by ngzhian** [comment]
The changes are here https://github.com/WebAssembly/spec/pull/1429

Basically I use the 'basic' theme from Sphinx, with custom pages (layout.html and page.html)
There's some JS kludge to mess around with the TOC to make the styling look more like the W3C toc.

**Comment by rossberg** [comment]
Btw, one issue with the current W3C doc is that it underlines all links even for single operators in math formulas, which can be really confusing and distracting. Is that still the case with this version? (The newhtml link seems dead, so I can't check.)

**Comment by ngzhian** [comment]
https://www.ngzhian.com/spec/newhtml/ is back up. (sorry my rebase messed up the published pages)

> Btw, one issue with the current W3C doc is that it underlines all links even for single operators in math formulas, which can be really confusing and distracting. Is that still the case with this version? (The newhtml link seems dead, so I can't check.)

It doesn't :)

**Comment by rossberg** [comment]
@ngzhian, @ericprud, what's the status of this issue, is it still relevant?

**Comment by ngzhian** [comment]
Not actively working on this.
